### DISCUSSION TITLE
The question of response header in remix

### DISCUSSION THREAD

**Discussion by 19Qingfeng** [remix-run/remix#6529] [Q&A] [answered]
I have some doubts and haven't found an answer on the document. I hope some friends can help me.

About the headers in LoaderFunction and the headers in HeaderFunction,I don't quite understand the usage scenarios of these two.

1. when i set this in my `root.tsx`:
```tsx
export const loader: LoaderFunction = async ({ request, context }) => {
  const headers = new Headers()
  headers.append('Set-Cookie','11111')
  headers.append('Cache-Control', 'no-store')
  return json(
    {},
    {
      headers,
    }
  )
}
```

After that, the responseHeader of set-cookie will exist in the returned html, but it's didn't have  cache-control header.

2. when i set this in my `root.tsx`:

```tsx
export const headers: HeaderFunction = async ({}) => {
  headers.append('Cache-Control', 'no-store')
  headers.append('Set-cookie', '123')
  return headers
}
```

Similarly, the Cache-Control response header does not exist in the returned html, but Set-cookie is working.

3. And Then, when i set this in my `entry.server.tsx`:

```tsx
function handleBrowserRequest(request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext) {
  // ...
  return new Promise(() => {
    // ...
     responseHeaders.set('Cache-Control', 'no-store')
     // ...
  })
}
```

The Cache-control response header takes effect at this time in my html.

I would like to ask what is the difference between these three.

I think what is written in `entry.server.tsx` will be to the whole for the html response.

But I'm not sure if there is any repetition of headers in loader and headers and their respective roles.

I tried to read the document and didn't figure out the difference, at least the above example seemed to confuse me. Can someone tell me the details of the similarities and differences between the three? thank you very much.

**Comment by lili21** [comment]
`Cache-Control` is controlled by leaf route. so setting it on Root wont' work.

**Comment by lili21** [comment, marked as answer]
The `Set-Cookie` header behaves differently in Remix. Whenever the loader or a parent route sets a `Set-Cookie` header, Remix will automatically preserve it. This is why the first scenario is successful.

However, for other headers, only the headers set by the leaf route will be applied, and you need to use the "headers" module to set them.

In the function "handleBrowserRequest", which outputs the HTML content, you can modify any header for a document request. But it is not recommended to make modifications within this function.
